stor/utils.py

```diff
@@ -229,6 +230,85 @@ def is_obs_path(p):
     return is_s3_path(p) or is_swift_path(p)
 
 
+def is_writeable(path):
+    """
+    Determine whether we have permission to write to path.
+
+    Behavior of this method is slightly different for different storage types when the
+    directory doesn't exist:
+    1. For local file systems, this function will return True if the target directory can
+       be created and a file written to it.
+    2. For AWS S3, this function will return True only if the target bucket is already
+       present and we have write access to the bucket.
+    3. For Swift, this function will return True, only if the target tenant is already
+       present and we have write access to the tenant and container. The container doesn't
+       have to be present.
+
+    This is function is useful, because `stor.stat()` will succeed if we have read-only
+    permissions to `path`, but the eventual attempt to upload will fail.
+
+    Secondly, `path` might not exist yet. If the intent of the caller is to create it, ,
+    stor.stat() will fail, however the eventual upload attempt would succeed.
+    """
+    from stor import basename
+    from stor import join
+    from stor import Path
+    from stor import remove
+    from stor import rmtree
+    from stor.swift import SwiftPath
+    from stor.swift import UnauthorizedError
+
+    path = Path(path)
+
+    # We want this function to behave as a no-op with regards to the underlying directory
+    # structure. Therefore we need to remove any directories created by this function that
+    # were not present when it was called. The `base_path_existed` defined below will
+    # store whether the directory that we're checking existed when calling this function,
+    # so that we know if it should be removed at the end.
+    if is_filesystem_path(path):
```

Comment:
>could we just use os.access for is_filesystem_path? That way we don't have to deal with the following situation:

I think it's the simplest approach. I could calculate what's the inner-most directory that exists before calling the function, but I think it makes sense for it to behave the same way as `osa

>This means additionally that the following doesn't work for filesystems:

`is_writeable(p)` returning true wasn't supposed to mean that you can immediately write files to `p`, but rather that you're allowed to create the directory structure described by `p` and then store files there. I think it makes sense to keep the behavior consistent with `os.access` in order not to confuse users who're used to `os.access` behavior.

Fixed in 773bafaa30d542f1df558ee7616b73f724b030ec